src/coreclr/jit/llvm.cpp

```diff
@@ -759,64 +834,196 @@ Value* buildCall(llvm::IRBuilder<>& builder, GenTree* node)
+    }
+    mapGenTreeToValue(node, builder.CreateCmp(llvmPredicate, op1, op2));
+}
+
+void buildPhi(llvm::IRBuilder<>& builder, GenTreePhi* phi)
+{
+    llvm::PHINode* llvmPhiNode = nullptr;
+    unsigned i = 0;
+    bool requiresLoad = false;
+    bool requiresLoadSet = false;
+    unsigned numChildren  = phi->NumChildren();
+    for (GenTreePhi::Use& use : phi->Uses())
+    {
+        GenTreePhiArg* phiArg = use.GetNode()->AsPhiArg();
+        unsigned       lclNum = phiArg->GetLclNum();
+        unsigned       ssaNum = phiArg->GetSsaNum();
+
+        LocatedLlvmValue localPhiArg = getSsaLocalForPhi(lclNum, ssaNum);
+        if (i == 0)
+        {
+            Type* phiType = getLLVMTypeForVarType(phi->TypeGet());
+
+            llvmPhiNode = builder.CreatePHI(requiresLoad ? phiType->getPointerTo() : phiType, numChildren);
+        }
+        if (localPhiArg.location != ValueLocation::ForwardReference)
+        {
+            if (requiresLoadSet)
+            {
+                // phi args must be all direct, or all indirect
+                assert(requiresLoad == localPhiArg.valueRequiresLoad());
+            }
+            else
+            {
+                requiresLoad    = localPhiArg.valueRequiresLoad();
+                requiresLoadSet = true;
+            }
+        }
+
+        if (localPhiArg.location == ValueLocation::ForwardReference)
+        {
+            _forwardReferencingPhis->insert({{lclNum, ssaNum}, {llvmPhiNode, getLLVMBasicBlockForBlock(phiArg->gtPredBB)}});
+        }
+        else
+        {
+            Value*             phiRealArgValue;
+            llvm::Instruction* castRequired = getCast(localPhiArg.getRawValue(), llvmPhiNode->getType());
+            if (castRequired != nullptr)
+            {
+                // This cast is needed when
+                // 1) The phi arg real type is short and the definition is the actual longer type, e.g. for bool/int
+                // 2) There is a pointer difference, e.g. i8* v i32* and perhaps different levels of indirection: i8** and i8*
+                llvm::BasicBlock::iterator phiInsertPoint = builder.GetInsertPoint();
+                llvm::BasicBlock* phiBlock = builder.GetInsertBlock();
+                llvm::Instruction* predBlockTerminator = getLLVMBasicBlockForBlock(phiArg->gtPredBB)->getTerminator();
+
+                builder.SetInsertPoint(predBlockTerminator);
+                phiRealArgValue = builder.Insert(castRequired);
+                builder.SetInsertPoint(phiBlock, phiInsertPoint);
+            }
+            else
+            {
+                phiRealArgValue = localPhiArg.getRawValue();
+            }
+            llvmPhiNode->addIncoming(phiRealArgValue, getLLVMBasicBlockForBlock(phiArg->gtPredBB));
+        }
+        i++;
+    }
+    assert(requiresLoadSet); // TODO-LLVM: If all the phi args are in this or later blocks, then this will be hit, and this load/no load swtich will have to be inserted when finishing the block, or something...
```

Comment:
Alternative solution to this problem is to turn all locals from the shadow stack into `IND`s in a pre-codegen pass (this will also be needed for "address exposed" locals).

Reply:
Thanks adding a link to this comment